### PR TITLE
[Validate] Add Metadata and Attribute filters to Metrics

### PR DESCRIPTION
**Pull request by gatli**
This builds on @Anirudh-Scale 3D Cuboid metrics PR: https://github.com/scaleapi/nucleus-python-client/pull/261

It hasn't been merged so the diff is huge 😅 

I've added `annotation_filters` and `prediction_filters` that allow arbitrary filtering of the data inputs to the metrics supporting `(cond AND cond AND cond ...) or (cond AND cond AND cond ...) or ...`. For a concrete example:

For example if you would like to test pedestrians closer than 70 meters with more than 15 points and pedestrians further away than 70 m with 5 points only against pedestrian annotations you could write something like this:
```
close_pedestrians = [MetadataFilter("relative_distance", "<", 70), [MetadataFilter("n_points", ">", 15), FieldFilter("label", "=", "pedestrian")]
farther_away_pedestrians = [MetadataFilter("relative_distance", ">", 70), [MetadataFilter("n_points", ">", 5), FieldFilter("label", "=", "pedestrian")]
either_close_or_farther = [close_pedestrians, farther_away_pedestrians]
only_pedestrian_annotations = [FieldFilter("label", "=" "pedestrian")]

evaluation_functions = [
        eval_functions.cuboid_precision(prediction_filters=[either_close_or_farther], annotation_filters=[only_pedestrian_annotations]),
        eval_functions.cuboid_recall(prediction_filters=[either_close_or_farther], annotation_filters=[only_pedestrian_annotations])]
        ]
 
config_description = [f"{f.key} {f.op} {f.value}" for f in close_pedestrians] + [[f"{f.key} {f.op} {f.value}" for f in farther_away_pedestrians]
test_name = f"Frog Lidar V2 - ({config_description})"
client.validate.create_scenario_test(test_name,
                                         training_slc_id, evaluation_functions=evaluation_functions)
```

The filters support the comparison operators:
```
class FilterOp(str, Enum):
    GT = ">"
    GTE = ">="
    LT = "<"
    LTE = "<="
    EQ = "="
    EQEQ = "=="
    NEQ = "!="
    IN = "in"
    NOT_IN = "not in"
```

## TODO:
- [x] Finish up adding the filters to all `Metric` classes
- [x] Make `shapely` installation optional

### CODE REVIEW

**Review by sasha-scale — COMMENTED** [review]
Overall code looks really well structured and clean - amazing work 🙌   

Testing locally, I think I found a small bug (although I wasn't able to root cause it).

Here's the bug report info:

https://dashboard.scale.com/nucleus/ut_c933c89spgb00ehnftjg?spoof=sasha.harrison@scale.com

Here's how I created the unit test:
```
from nucleus.metrics import MetadataFilter, FieldFilter
BDD_TEST_SLICE = 'slc_c933az3ptmpg0cs2vyk0'
example_annotation_filter = [MetadataFilter("occluded", "==", False), FieldFilter("label", "=", "pedestrian")]
evaluation_functions = [
        client.validate.eval_functions.cuboid_precision(annotation_filters=example_annotation_filter),
        client.validate.eval_functions.cuboid_recall(annotation_filters=example_annotation_filter)
]
config_description = [f"{f.key} {f.op} {f.value}" for f in example_annotation_filter]
test_name = f"BDD Experiments 1 - ({config_description})"
test = client.validate.create_scenario_test(test_name,
                                         BDD_TEST_SLICE, evaluation_functions=evaluation_functions)
```

The behavior I observed was that when I evaluated the unit test, there was an exception on the ML side:  the error is:

```
"Traceback (most recent call last):\n File \"/usr/local/lib/python3.8/site-packages/celery/app/trace.py\", line 405, in trace_task\n R = retval = fun(*args, **kwargs)\n File \"/usr/local/lib/python3.8/site-packages/celery/app/trace.py\", line 697, in __protected_call__\n return self.run(*args, **kwargs)\n File \"/workspace/model_ci/services/evaluate/tasks.py\", line 51, in evaluate\n response: MetricEvaluationResponse = evaluate_metric(request)\n File \"/workspace/model_ci/services/metrics.py\", line 66, in evaluate_metric\n per_item_metrics, overall_metric = compute_metric(preds_grouped, gt_grouped, eval_func)\n File \"/workspace/model_ci/services/metrics.py\", line 42, in compute_metric\n result = nucleus_metric(anno_list, pred_list)\n File \"/usr/local/lib/python3.8/site-packages/nucleus/metrics/cuboid_metrics.py\", line 272, in __call__\n cuboid_annotations = filter_by_metadata_fields(\n File \"/usr/local/lib/python3.8/site-packages/nucleus/metrics/cuboid_metrics.py\", line 190, in filter_by_metadata_fields\n and_conditions = [\n File \"/usr/local/lib/python3.8/site-packages/nucleus/metrics/cuboid_metrics.py\", line 191, in <listcomp>\n filter_to_comparison_function(cond) for cond in or_branch\n File \"/usr/local/lib/python3.8/site-packages/nucleus/metrics/cuboid_metrics.py\", line 127, in filter_to_comparison_function\n if FilterType(metadata_filter.type) == FilterType.FIELD:\nAttributeError: 'str' object has no attribute 'type'\n"
```

**Comment by gatli** [comment]
> Overall code looks really well structured and clean - amazing work 🙌
> 
> Testing locally, I think I found a small bug (although I wasn't able to root cause it).
> 
> Here's the bug report info:
> 
> https://dashboard.scale.com/nucleus/ut_c933c89spgb00ehnftjg?spoof=sasha.harrison@scale.com
> 
> Here's how I created the unit test:
> 
> ```
> from nucleus.metrics import MetadataFilter, FieldFilter
> BDD_TEST_SLICE = 'slc_c933az3ptmpg0cs2vyk0'
> example_annotation_filter = [MetadataFilter("occluded", "==", False), FieldFilter("label", "=", "pedestrian")]
> evaluation_functions = [
>         client.validate.eval_functions.cuboid_precision(annotation_filters=example_annotation_filter),
>         client.validate.eval_functions.cuboid_recall(annotation_filters=example_annotation_filter)
> ]
> config_description = [f"{f.key} {f.op} {f.value}" for f in example_annotation_filter]
> test_name = f"BDD Experiments 1 - ({config_description})"
> test = client.validate.create_scenario_test(test_name,
>                                          BDD_TEST_SLICE, evaluation_functions=evaluation_functions)
> ```
> 
> The behavior I observed was that when I evaluated the unit test, there was an exception on the ML side: the error is:
> 
> ```
> "Traceback (most recent call last):\n File \"/usr/local/lib/python3.8/site-packages/celery/app/trace.py\", line 405, in trace_task\n R = retval = fun(*args, **kwargs)\n File \"/usr/local/lib/python3.8/site-packages/celery/app/trace.py\", line 697, in __protected_call__\n return self.run(*args, **kwargs)\n File \"/workspace/model_ci/services/evaluate/tasks.py\", line 51, in evaluate\n response: MetricEvaluationResponse = evaluate_metric(request)\n File \"/workspace/model_ci/services/metrics.py\", line 66, in evaluate_metric\n per_item_metrics, overall_metric = compute_metric(preds_grouped, gt_grouped, eval_func)\n File \"/workspace/model_ci/services/metrics.py\", line 42, in compute_metric\n result = nucleus_metric(anno_list, pred_list)\n File \"/usr/local/lib/python3.8/site-packages/nucleus/metrics/cuboid_metrics.py\", line 272, in __call__\n cuboid_annotations = filter_by_metadata_fields(\n File \"/usr/local/lib/python3.8/site-packages/nucleus/metrics/cuboid_metrics.py\", line 190, in filter_by_metadata_fields\n and_conditions = [\n File \"/usr/local/lib/python3.8/site-packages/nucleus/metrics/cuboid_metrics.py\", line 191, in <listcomp>\n filter_to_comparison_function(cond) for cond in or_branch\n File \"/usr/local/lib/python3.8/site-packages/nucleus/metrics/cuboid_metrics.py\", line 127, in filter_to_comparison_function\n if FilterType(metadata_filter.type) == FilterType.FIELD:\nAttributeError: 'str' object has no attribute 'type'\n"
> ```

I'll have a look. I actually just fixed list of `AND` re-parsing, I think it might be related to that. Thanks for testing it end-to-end!

File: nucleus/validate/client.py [hunk 1]
```diff
@@ -83,7 +86,10 @@ def create_scenario_test(
                 name=name,
                 slice_id=slice_id,
                 evaluation_functions=[
-                    ef.to_entry() for ef in evaluation_functions  # type:ignore
+                    EvalFunctionListEntry(
```
[the remark this answers is not in the document]
Comment (gatli) [reply nucleus/validate/client.py:89]: No problem! Your solution was fine but I just disliked that it made some immutable fields look mutable 🙂

**Comment by gatli** [comment]
I've updated the deployment and tested all public functions. They all work in my tests 🙂

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,15 @@ All notable changes to the [Nucleus Python Client](https://github.com/scaleapi/n
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.8.4](https://github.com/scaleapi/nucleus-python-client/releases/tag/v0.8.3) - 2022-04-05
+
+### Added
+
+- Validate metrics support metadata and field filtering on input annotation and predictions
+- 3D Cuboid Recall, Precision and IOU metrics
```
Comment (pfmark) [inline CHANGELOG.md:12]:
```suggestion
- 3D/Cuboid metrics: Recall, Precision, 3D IOU and birds eye 2D IOU```

File: README.md [hunk 3]
```diff
@@ -179,3 +179,22 @@ cd docs
...
+Note that you might need to install a local GEOS package.
+
+```bash
+#Mac OS
+brew install geos
+# Ubuntu/Debian flavors
+apt-get install libgeos-dev
+```
+
+`pip install scale-nucleus[metrics]`
```
Comment (pfmark) [inline README.md:196]: The default installation would still be `pip install scale-nucleus`?

File: nucleus/metrics/filtering.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+)
+
+
+class FilterOp(str, Enum):
+    GT = ">"
+    GTE = ">="
+    LT = "<"
+    LTE = "<="
+    EQ = "="
+    EQEQ = "=="
```
Comment (pfmark) [inline nucleus/metrics/filtering.py:32]: can we just put `==` here? They are both mapped to `==` anyways. I think having both of them is confusing.

**Review by pfmark — COMMENTED** [review]
overall really great work! 🚀 

Tested it locally and I have a few questions:
- I added pretty arbitrary filters (`[MetadataFilter("relative_distance", ">", 50000), FieldFilter("label", "=", "lalala")]`) which should result in no items being passed to the test. Evaluating the [test](https://dashboard.scale.com/nucleus/ut_c96prrkhe03g0b1hskhg?spoof=mark.pfeiffer@scale.com) with a model that has the metadata and a model that doesn't have the relevant metadata still has results. Any thought why this happens? Am I getting sth. wrong here?
- Any chance we can add some feedback to the scenario test creation which would let the user know how many items were filtered per evaluation function? I think this would be super useful.

File: README.md (hunk 3 above)
Comment (gatli) [reply README.md:196]: Yup! But this should be `pip install scale-nucleus[shapely]`

File: nucleus/metrics/filtering.py (hunk 4 above)
Comment (gatli) [reply nucleus/metrics/filtering.py:32]: One is Python style and the other is SQL style, I think it is fine practice to support both, PyArrow does the same. It has no effect on maintainability but might save somebody a headache so I'll keep it in. 🙂

**Pull request merged**